AI #1121: Rename `conformance` to `requirements model` (#1526)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/requirements_model/README.md

> # Requirements Model (RM)

Reviewer: ```suggestion
# FOCUS Requirements Model (RM)
```

> Requirements Model are structured, testable rules that define how each column in a dataset must behave in order to comply with the FOCUS specification. These rules are extracted from **verbose normative text** and translated into precise **logic statements** that can be programmatically validated. Each requirement includes elements such as _applicability conditions_, _expected behavior_, and _dependencies_ on other rules—enabling both human-readable documentation and automated conformance checking.

Reviewer: ```suggestion
The FOCUS Requirements Model is a structured, testable set of rules that define how each column in a dataset must behave in order to comply with the FOCUS specification. These rules are extracted from **verbose normative text** in the FOCUS specification and translated into precise **logic statements** that can be programmatically validated. Each requirement includes elements such as _applicability conditions_, _expected behavior_, and _dependencies_ on other rules—enabling both human-readable documentation and automated conformance checking.
```

## Introduction

> This document outlines a structured approach to converting verbose technical requirements from the FOCUS v1.2 specification into precise, programmatically testable Requirements Model. The goal is to enable consistent validation, documentation, and visualization of rules across all dataset columns using structured JSON representations.

Reviewer: ```suggestion
This document outlines a structured approach to converting verbose technical requirements from the FOCUS v1.2 specification into precise, programmatically testable rules. The goal is to enable consistent validation, documentation, and visualization of rules across all dataset columns using structured JSON representations.
```

The project is divided into three main stages, each with a defined scope and estimated effort:

> ### Stage 1 – Extract Requirements Model from FOCUS v1.2  

Reviewer: ```suggestion
### Stage 1 – Extract Normative Requirements from FOCUS v1.2  
```

**Estimated Effort: 40% of total project time**

> This foundational stage involves analyzing the **FOCUS v1.2** Technical Specification to extract **atomic** and **composite**  Model Rule s for each dataset column. These are captured in standardized Markdown tables that define logic, applicability, conditions, and dependencies.

Reviewer: ```suggestion
This foundational stage involves analyzing the **FOCUS v1.2** Technical Specification to extract **atomic** and **composite**  normative requirements for each attribute and dataset column. These are captured in standardized Markdown tables that define logic, applicability, conditions, and dependencies.
```

The JSON output enables programmatic validation and integration into testing workflows, making it a crucial bridge between human-readable requirements and tooling.

### Stage 3 – Model Dependencies Between ModelIds  

**Estimated Effort: 25% of total project time**

This stage focuses on interpreting the `Requirement` field of each **RuleId** to explicitly map dependencies to other **ModelIds** using logical structures such as `AND`, `OR`, and `NOT`. The output is used to construct a dependency graph for each column, enabling accurate validation of composite rules.

**Stage 3** can only be properly completed once both **Stage 1** and **Stage 2** are finalized, as it depends on having both accurate ModelId identifiers and their JSON representations available for analysis.

These three stages form the foundation of the modeling workflow. Once complete, additional outputs will be generated from the JSON representations, including:

> - A standardized Markdown table of Requirements Model for each column

Reviewer: ```suggestion
- A standardized Markdown table of normative requirements for each column
```

---

> ## 🧾 model Expression Format

Reviewer: ```suggestion
## 🧾 Model Expression Format
```

## 📐 RFC 2119 Keyword Mapping

> | Keyword                | Default model Classification | Notes                                      |

Reviewer: ```suggestion
| Keyword                | Default Model Classification | Notes                                      |
```

## Versioning

> In general, the version of each Requirement Model (RM) is aligned with the version of the FOCUS Specification to which it applies. However, there may be occasions where a correction or clarification is required due to an error or bug in a previously published model definition.

Reviewer: ```suggestion
In general, the version of each FOCUS Requirement Model (RM) is aligned with the version of the FOCUS Specification to which it applies. However, there may be occasions where a correction or clarification is required due to an error or bug in a previously published model definition.
```

> To accommodate these changes without altering the core FOCUS Specification version, the `Version` field of a model may differ slightly from the associated FOCUS Specification version. This allows us to make targeted corrections while preserving compatibility and traceability.

Reviewer: ```suggestion
To accommodate these changes without altering the core FOCUS Specification version, the `Version` field of a FOCUS Requirements Model may differ slightly from the associated FOCUS Specification version. This allows us to make targeted corrections while preserving compatibility and traceability.
```
